File: gek-core/src/main/java/xyz/fslabo/common/io/BytesOutputStream.java
package xyz.fslabo.common.io;

import java.io.IOException;
import java.io.OutputStream;

final class BytesOutputStream extends OutputStream {

    private final byte[] buf;
    private final int end;
    private int pos;

    BytesOutputStream(byte[] buf) {
        this(buf, 0, buf.length);
    }

    BytesOutputStream(byte[] buf, int offset, int length) {
        IOMisc.checkReadBounds(buf, offset, length);
        this.buf = buf;
        this.end = offset + length;
        this.pos = offset;
    }

    @Override
    public void write(int b) throws IOException {
        if (end - pos < 1) {
            throw new IOException("The backing array has insufficient capacity remaining.");
        }
        buf[pos] = (byte) b;
        pos++;
    }

    @Override
    public void write(byte[] b, int off, int len) throws IOException {
        IOMisc.checkReadBounds(b, off, len);
        if (len <= 0) {
            return;
        }
        if (end - pos < len) {
            throw new IOException("The backing array has insufficient capacity remaining.");
        }
        System.arraycopy(b, off, buf, pos, len);
        pos += len;
    }
}

File: boat-core/src/main/kotlin/xyz/srclab/common/io/BytesOutputStream.kt
package xyz.srclab.common.io

import xyz.srclab.common.base.checkInBounds
import xyz.srclab.common.base.checkRangeInBounds
import java.io.OutputStream

/**
 * Makes array as destination of [OutputStream].
 */
open class BytesOutputStream(
    private val destination: ByteArray,
    private val offset: Int,
    private val length: Int
) : OutputStream() {

    init {
        checkRangeInBounds(offset, offset + length, 0, destination.size)
    }

    private var pos = offset

    override fun write(b: Int) {
        pos.checkInBounds(offset, offset + length)
        destination[pos] = b.toByte()
        pos++
    }

    override fun write(b: ByteArray, off: Int, len: Int) {
        checkRangeInBounds(off, off + len, 0, b.size)
        val nextPos = pos + len
        (nextPos - 1).checkInBounds(offset, offset + length)
        System.arraycopy(b, off, destination, pos, len)
        pos = nextPos
    }
}